### ISSUE TITLE
Feature Suggestion: ignore hard linked files by default

### ISSUE THREAD

**Issue by halfmanhalffish** [beyondgrep/ack3#314]
Since hard linked files are identical, search results from them will always be duplicates. It'd be a nice power tool feature to ignore all but one of them by default. Eg:
ls
file1
ln file1 subdir/file1
stat file1
[snip]
Device: 802h/2050d      Inode: 6325564     Links: 2
ack searchterm
./file1: ...searchterm...
[but no search results from subdir/file1]

**Comment by hoelzro** [comment]
ack already has `get_file_id`, which it uses to skip over duplicates specified on the command line, so this probably wouldn't be too bad to add!

**Comment by n1vux** [comment]
Since we already have the tooling to skip duplicates (whether by fs+inode or hash), having an **option** to omit duplicates seems very reasonable. 

OTOH i frequently do want to know the duplicate files also match --  especially with `-l` (which files not which lines) -- so this canot be the default behavior with `-l` at least. 

On forward compatibility grounds, I suggest this not be the default behavior (in ack3.$n++) but users who wish it to be default behavior can affirmatively set the needed option in .ackrc. (a hypothetical ack4 can change defaults incompatibly, of course).

How does this handle soft-link aliases ? The final resolved dev:inode will still be the same so those would be weeded out too?

**Comment by petdance** [comment]
Before we get into the "how it could be done", let's talk interface.

In your example where you have, say:

    ln foo foo1
    ln foo subdir/foo2
    ln foo other/dir/foo3
    ln foo4

Which file gets recognized and which ones get ignored?

Will it just be whichever one gets directory traversed first?  That means that sometimes you might get a hit in `foo` and the next time it's in `foo4` because we don't know what order the entries are coming back int.

**Comment by n1vux** [comment]
There are several reasons this should not be "**by default**"

- Memory usage (potentially big uptick on large filesystem) (h/t @hoelzro )
- backwards compatibility
- grep compatibility
- hiding information by default
- users such as OP who want this default have an easy way to do that, in .ackrc

**Comment by n1vux** [comment]
> Which file gets recognized and which ones get ignored?
> Will it just be whichever one gets directory traversed first?

With any known Unix/Linux filesystem, there is no memory of which hardlink was "first", so **yes**. 

If the order files are reported by Ack is non-deterministic, which will be suppressed/included will be also under this option. 

We already have an option `--sort-files` to pay extra to get determinism, so that is an already solved problem?

**Comment by n1vux** [comment]
> Before we get into the "how it could be done", let's talk interface.

I suspect point of @hoelzro 's "implementation" comment is that we're _already_ paying the filesystem hit performance penalty needed for this, so it's a "_low hanging fruit_" as opposed to "_expensive wishlist but don't hold your breath_". 

(Rob is correct that on a large filesystem there could be significant memory cost to keeping an additional hash, even if it's just `{"$dev:$inode"=>1 ,...}` but if it's a cost only paid on an affirmative option that doesn't bother me.)

**Comment by n1vux** [comment]
> With any known Unix/Linux filesystem, there is no memory of which hardlink was "first", so **yes**.

Just because i don't know of one doesn't mean there is not a single one ... but that's all we can assume (and not even that on Windows!)

Does MacOSX f/s remember which hardlink was original primary? 
They have extra metadata of strange sorts ...

**Comment by petdance** [comment]
So the question, @halfmanhalffish, is: If you can get different results from two different runs of `ack` without any of the contents of your files changing, is that still a useful feature?

**Comment by halfmanhalffish** [comment]
Sorry @petdance I don't understand your meaning. 'two different runs' ... do you mean 2 runs with different switches?

Incidentally, I've worked around this hardlink problem by using ag (apologies!!! but ack doesn't seem to have this feature) which by default ignores files listed in .gitignore (which naturally includes the hard linked files I want to disregard).

**Comment by petdance** [comment]
> Sorry @petdance I don't understand your meaning. 'two different runs' ... do you mean 2 runs with different switches?

No.  I'm saying that if we ignore hard links it is possible that you could get two different sets of results from the same invocation of ack, without the contents of the files changing.

 ```
$ echo blah > foo
$ ln foo foo1
$ ln foo foo2
$ ln foo foo3
$ ack blah
foo:1:blah
$ ack blah
foo2:1:blah
```

Note how the first invocation of ack reported a match in `foo`, and the second in `foo2`, even though we didn't do anything to change the file contents.  That's because we don't know what order we'll be getting filenames in.

>  by using ag (apologies!!! but ack doesn't seem to have this feature)

No reason to apologize.  [I want you to use the tool that works best for you.](https://blog.petdance.com/2018/01/02/the-best-open-source-project-for-someone-might-not-be-yours-and-thats-ok/)  I'm pretty sure that ripgrep does the same thing.  Here's a [feature comparison chart](https://beyondgrep.com/feature-comparison/).

You can also put some `--ignore-file` options in your project's `.ackrc`.  That would achieve the same thing, although it would mean having the same files ignored in two different ways in two different files (.gitignore and .ackrc).

**Comment by halfmanhalffish** [comment]
If ack is invoked the same way and the fs isn't changed then one wouldn't expect different output. But you're right, if certain ack switches meant the fs was being searched differently then one wouldn't want matches from different hard linked files showing up.

Someone wrote above that this feature should be opt-in (and I think he was right). So if there's to be config or switches then why not just require the config to specify which hard linked file to check?

**Comment by petdance** [comment]
> If ack is invoked the same way and the fs isn't changed then one wouldn't expect different output. 

Right, but who knows how or when the filesystem will change?

> why not just require the config to specify which hard linked file to check?

I'm not sure what you mean.  If you've got `foo`, `foo2` and `foo3` that are all hard links to each other, how would we specify which one of the three is the one that you'd want to have matched, and which two would be ignored?

**Comment by halfmanhalffish** [comment]
Perhaps ...

ack --ignore-hard-linked-files
would show matches only from the file 'nearest' to the working directory, where nearest would be judged either by file+pathname length as a string, or by the number of subdirectories.

Or specify files explicitly
ack --ignore-hard-linked-files=dir/file1,dir/dir2/file2

**Comment by petdance** [comment]
> ack --ignore-hard-linked-files would show matches only from the file 'nearest' to the working directory

But what if they're all in the same directory, as in my example above?

> ack --ignore-hard-linked-files=dir/file1,dir/dir2/file2

You already have `--ignore-file=is:dir/file1`.

**Comment by halfmanhalffish** [comment]
If all in same directory then pick the one with the shortest name. If same name length then lexically in the strcmp() sense, ie 'a' < 'b'.
Good point about --ignore-file=is:dir/file1.
